### PR TITLE
core: ffa: Improve FF-A memory sharing compliance

### PR DESCRIPTION
**Pull request by imre-kis-arm**
* Deny memory regions with zero pages
* Validate total page count field
* Validate total descriptor size including memory regions descriptors
* Deny unsupported fragmented operations
* Fix incorrect FFA_ERROR status codes
* Return transaction type flag in retrieve response

Added test cases to SPMC test SPs to cover error handling branches in the memory sharing related functions: https://review.trustedfirmware.org/c/TS/trusted-services/+/28384

### CODE REVIEW

**Review by jenswi-linaro — COMMENTED** [review]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/kernel/thread_spmc.c [hunk 1]
```diff
@@ -1184,13 +1187,24 @@ static int handle_mem_share_tmem(paddr_t pbuf, size_t blen, size_t flen,
...
+	if (rc)
+		goto unlock;
+
+	if (is_sp_share(&mem_trans, buf)) {
+		rc = spmc_sp_add_share(&mem_trans, buf, blen, flen,
+				       global_handle, NULL);
+		goto unlock;
+	}
+
+	if (!IS_ENABLED(CFG_NS_VIRTUALIZATION) &&
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread_spmc.c:1199]: Shouldn't this be `if (IS_ENABLED(CFG_NS_VIRTUALIZATION) && virt_set_guest(mem_trans.sender_id))` ?

**Review by etienne-lms — COMMENTED** [review]
Many changes in this commit, a bit hard to follow. Maybe split it in pieces? That is just a suggestion, not a very strong opinion.

File: core/arch/arm/kernel/thread_spmc.c (hunk 1 above)
Comment (imre-kis-arm) [reply core/arch/arm/kernel/thread_spmc.c:1199]: You're right, I fixed this line.

**Review by etienne-lms — COMMENTED** [review]
2 nitpicking comments, otherwise looks good to me.
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for the series, with the my comments addressed or not.

File: core/arch/arm/kernel/thread_spmc.c [hunk 2]
```diff
@@ -1184,13 +1184,24 @@ static int handle_mem_share_tmem(paddr_t pbuf, size_t blen, size_t flen,
...
+	}
+
+	if (IS_ENABLED(CFG_NS_VIRTUALIZATION) &&
 	    virt_set_guest(mem_trans.sender_id))
 		rc = FFA_DENIED;
 	if (!rc)
 		rc = add_mem_share(&mem_trans, mm, buf, blen, flen,
 				   global_handle);
 	virt_unset_guest();
+
```
[the remark this answers is not in the document]
Comment (imre-kis-arm) [reply core/arch/arm/kernel/thread_spmc.c:1206]: Ok, I changed this part and added `IS_ENABLED(CFG_NS_VIRTUALIZATION)` condition to all `virt_unset_guest()` calls in this file.

File: core/arch/arm/kernel/spmc_sp_handler.c [hunk 3]
```diff
@@ -297,9 +297,14 @@ int spmc_sp_add_share(struct ffa_mem_transaction_x *mem_trans,
 	size_t addr_range_offs = 0;
 	struct ffa_mem_region *mem_reg = NULL;
 	uint8_t highest_permission = 0;
-	struct sp_mem *smem = sp_mem_new();
+	struct sp_mem *smem = NULL;
 	uint16_t sender_id = mem_trans->sender_id;
 
+	/* Fragmented memory share is not supported for SPs */
+	if (blen != flen)
+		return FFA_NOT_SUPPORTED;
```
[the remark this answers is not in the document]
Comment (imre-kis-arm) [reply core/arch/arm/kernel/spmc_sp_handler.c:309]: Converted the comment into a `DMSG()`

**Comment by imre-kis-arm** [comment]
> 2 nitpicking comments, otherwise looks good to me. `Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for the series, with the my comments addressed or not.

Added Acked-by.

**Comment by imre-kis-arm** [comment]
I rebased the PR so it has the `FORCE_UNSAFE_CONFIGURE` fix but now it fails when it tries to download the Rust toolchain. Is this a known issue or is it just a temporary network error?

**Comment by imre-kis-arm** [comment]
Could you please merge this PR if there are no further concerns with it?

**Comment by jforissier** [comment]
@imre-kis-arm sure, sorry I missed it.

**Comment by imre-kis-arm** [comment]
Thank you.
